src/main/kotlin/org/rust/ide/hints/RsGenericParameterInfoHandler.kt

```diff
@@ -0,0 +1,148 @@
+    }
+    
+    // todo: don't disappear in nested generic types
+    override fun updateParameterInfo(parameterOwner: RsTypeArgumentList, context: UpdateParameterInfoContext) {
+        curParam = ParameterInfoUtils.getCurrentParameterIndex(parameterOwner.node, context.offset, RsElementTypes.COMMA)
+        when {
+            context.parameterOwner == null -> context.parameterOwner = parameterOwner
+
+            // occurs e.g. in case of up-down cursor moving
+            context.parameterOwner != parameterOwner -> context.removeHint()
+        }
+    }
+
+    override fun updateUI(p: RsGenericPresentation, context: ParameterInfoUIContext) {
+        hintText = p.presentText
+        val (startOffset, endOffset) =
+            if (hintText.startsWith(WHERE_PREFIX))
+                0 to WHERE_PREFIX.length
+            else
+                p.getRange(curParam).startOffset to p.getRange(curParam).endOffset
+        context.setupUIComponentPresentation(
+            hintText,
+            startOffset,
+            endOffset,
+            false, // grayed
+            false,
+            false, // define grayed part of args before highlight
+            context.defaultParameterColor
+        )
+    }
+
+    override fun getParametersForLookup(item: LookupElement?, context: ParameterInfoContext?): Array<Any>? = null
+
+    override fun couldShowInLookup() = false
+
+    override fun findElementForUpdatingParameterInfo(context: UpdateParameterInfoContext): RsTypeArgumentList? =
+        findExceptColonColon(context)
+
+    override fun findElementForParameterInfo(context: CreateParameterInfoContext): RsTypeArgumentList? {
+        val parameterList = findExceptColonColon(context) ?: return null
+        val parent = parameterList.parent
+        val genericDeclaration = when (parent) {
+            is RsMethodCall,
+            is RsPath -> parent.reference?.resolve()
+            else -> return null
+        } as? RsGenericDeclaration ?: return null
+        val typesWithBounds = genericDeclaration.typeParameters.nullize() ?: return null
+        context.itemsToShow = listOfNotNull(
+            RsGenericPresentation(typesWithBounds, false),
+            RsGenericPresentation(typesWithBounds, true))
+            .filterNot { it.presentText == "" }
+            .toTypedArray()
+        return parameterList
+    }
+
+    // to avoid hint on :: before <>
+    private fun findExceptColonColon(context: ParameterInfoContext?): RsTypeArgumentList? {
+        val element = context?.file?.findElementAt(context.editor.caretModel.offset) ?: return null
+        if (element.elementType == RsElementTypes.COLONCOLON) return null
+        return element.ancestorStrict() ?: return null
+    }
+}
+
+/**
+ * Stores the text representation and ranges for parameters
+ */
+
+class RsGenericPresentation(
+    private val params: List<RsTypeParameter>,
+    unrecognizedPartOfWhere: Boolean
+) {
+    val toText = if (!unrecognizedPartOfWhere) {
```

Comment:
This is rather complicated code; consider splitting it into several functions

Reply:
Can be private.
And let's specify types explicitly for each property.
Also, the name `toText` looks like a function name because it's a verb (_convert_ to text). Even more confusing is the fact that this is a **field**, not a kotlin property. Maybe really convert it to a function and compute the value lazily?